File: backend/api/routes/hla.py
"""HLA / HIBAG API (Wave D).

- ``GET /api/hla/status`` (SW-D1): whether the operator-installed HIBAG HLA-
  imputation runtime is usable (``Rscript`` present + a BYO ancestry model), so the
  UI can choose between full HIBAG calls and the single-tag HLA proxy fallback.
  Runtime/config status, not sample-specific.
- ``GET /api/hla/drug-hypersensitivity?sample_id=N`` (SW-D2): per-drug
  hypersensitivity-risk assessments from a sample's imputed classical-HLA calls
  (well-established CPIC-grade HLA–drug contraindications). Imputed, not typed —
  every response carries the confirmatory-typing caveat.
"""

from __future__ import annotations

import sqlalchemy as sa
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from backend.analysis.hibag_runner import hibag_runtime_status
from backend.analysis.hla_drug_hypersensitivity import assess_drug_hypersensitivity
from backend.analysis.hla_resolver import read_hla_calls
from backend.db.connection import get_registry
from backend.db.tables import samples

router = APIRouter(prefix="/hla", tags=["hla"])


class HibagStatusResponse(BaseModel):
    """HIBAG runtime availability."""

    rscript_available: bool
    model_dir_configured: bool
    ancestry_models: list[str]
    available: bool


@router.get("/status", response_model=HibagStatusResponse)
async def get_hla_status() -> HibagStatusResponse:
    """Report HIBAG HLA-engine availability (Rscript + BYO ancestry models)."""
    settings = get_registry().settings
    status = hibag_runtime_status(settings.hibag_rscript, settings.hibag_model_dir)
    return HibagStatusResponse(
        rscript_available=status.rscript_available,
        model_dir_configured=status.model_dir is not None,
        ancestry_models=status.ancestry_models,
        available=status.available,
    )


def _get_sample_engine(sample_id: int) -> sa.Engine:
    registry = get_registry()
    with registry.reference_engine.connect() as conn:
        row = conn.execute(
            sa.select(samples.c.db_path).where(samples.c.id == sample_id)
        ).fetchone()
    if row is None:
        raise HTTPException(status_code=404, detail=f"Sample {sample_id} not found.")
    sample_db_path = registry.settings.data_dir / row.db_path
    if not sample_db_path.exists():
        raise HTTPException(
            status_code=404, detail=f"Sample database file not found for sample {sample_id}."
        )
    return registry.get_sample_engine(sample_db_path)


class DrugRiskAssessmentResponse(BaseModel):
    """The sample's assessment against one HLA drug-hypersensitivity association."""

    allele: str
    drugs: list[str]
    reaction: str
    status: str  # at_risk | no_risk_allele | not_typed
    carried: bool
    zygosity: str | None
    copies: int
    prob: float | None
    low_confidence: bool
    recommendation: str
    guideline: str
    citations: list[str]
    notes: list[str]


class DrugHypersensitivityResponse(BaseModel):
    """HLA drug-hypersensitivity report for a sample."""

    available: bool
    any_at_risk: bool
    assessments: list[DrugRiskAssessmentResponse] = []
    caveat: str = ""
    unavailable_note: str | None = None
    research_use_only: bool = True


@router.get("/drug-hypersensitivity", response_model=DrugHypersensitivityResponse)
def get_hla_drug_hypersensitivity(
    sample_id: int = Query(..., description="Sample ID"),
) -> DrugHypersensitivityResponse:
    """Assess a sample's imputed HLA calls for drug-hypersensitivity contraindications."""
    engine = _get_sample_engine(sample_id)
    report = assess_drug_hypersensitivity(read_hla_calls(engine))
    return DrugHypersensitivityResponse(
        available=report.available,
        any_at_risk=report.any_at_risk,
        assessments=[DrugRiskAssessmentResponse(**vars(a)) for a in report.assessments],
        caveat=report.caveat,
        unavailable_note=report.unavailable_note,
    )

File: frontend/src/types/hla.ts
/** HLA (imputed) module API types (Wave D). */

export type HlaDrugRiskStatus = "at_risk" | "low_confidence" | "no_risk_allele" | "not_typed"

export interface HlaDrugRiskAssessment {
  allele: string
  drugs: string[]
  reaction: string
  status: HlaDrugRiskStatus
  carried: boolean
  zygosity: string | null
  copies: number
  prob: number | null
  low_confidence: boolean
  recommendation: string
  guideline: string
  citations: string[]
  notes: string[]
}

export interface HlaDrugHypersensitivityResponse {
  available: boolean
  any_at_risk: boolean
  assessments: HlaDrugRiskAssessment[]
  caveat: string
  unavailable_note: string | null
  research_use_only: boolean
}

type CeliacRuleOutStatus = "rule_out" | "permissive_present" | "not_typed"
type NarcolepsyRuleOutStatus = "absent_lowers" | "present" | "not_typed"

export interface CeliacRuleOut {
  status: CeliacRuleOutStatus
  detected: string[]
  low_confidence: boolean
  interpretation: string
}

export interface NarcolepsyRuleOut {
  status: NarcolepsyRuleOutStatus
  carried: boolean
  zygosity: string | null
  low_confidence: boolean
  interpretation: string
}

export interface HlaRuleOutsResponse {
  available: boolean
  celiac: CeliacRuleOut | null
  narcolepsy: NarcolepsyRuleOut | null
  caveat: string
  unavailable_note: string | null
  citations: string[]
  research_use_only: boolean
}

export type HlaSusceptibilityStatus =
  | "increased_risk"
  | "not_increased"
  | "neutral_subtype"
  | "not_typed"

export interface HlaSusceptibilityFinding {
  condition: string
  hla: string
  status: HlaSusceptibilityStatus
  carried: boolean
  detail: string
  interpretation: string
  low_confidence: boolean
  citations: string[]
  notes: string[]
}

export interface HlaSusceptibilityResponse {
  available: boolean
  findings: HlaSusceptibilityFinding[]
  caveat: string
  unavailable_note: string | null
  research_use_only: boolean
}

/** One imputed per-locus HLA genotype row. @public — element type of HlaViewerResponse.alleles, for downstream consumers. */
export interface HlaAlleleView {
  locus: string
  allele1: string
  allele2: string
  prob: number | null
  low_confidence: boolean
  source: string
  ancestry_model: string | null
}

export interface HlaViewerResponse {
  available: boolean
  alleles: HlaAlleleView[]
  caveat: string
  transplant_guard: string
  unavailable_note: string | null
  research_use_only: boolean
}
